### PR TITLE
Use unpacking inside arrays

### PR DESCRIPTION
**Pull request by AJenbo**
Type: refactoring
Breaking change: no

With unpacking in arrays (introduced in PHP 7.4) `array_merge()` can now be replaced by operators.

### CODE REVIEW

**Comment by ravage84** [comment]
Hm... what is the actual goal of this one? Or what is the problem with `array_merge`?

Have you considered the varying behavior of both ways?

https://stitcher.io/blog/what-is-array-plus-in-php

https://www.php.net/manual/en/function.array-merge.php
https://www.php.net/manual/en/language.operators.array.php


Also, if [this comment](https://www.php.net/manual/en/function.array-merge.php#126687) is true, it might even be a minor performance downgrade.

🤔

**Comment by AJenbo** [comment]
> Hm... what is the actual goal of this one? Or what is the problem with array_merge?

I find the new syntax cleaner. Same with array union. Sadly none of them could be reduced to `+=` but still:
```php
$items = $proportions + $metrics;
```
Then the `array_merge()` variant:
```php
$items = array_merge($metrics, $proportions);`
```

For the splat operators I particularly find this one a big improvment:
```php
return ['metadata', ...parent::__sleep()];
```
over `array_merge()` since it doesn't create an array just to satisfy an interface:
```php
return array_merge(['metadata'], parent::__sleep());
```

In PHPMD you will also see some where it is nested in other calls which can make it a bit harder to see what is happening when using array_merge:

```php
$args = array_unique(array_filter(array_merge([__CLASS__, $this::class], $this->getClasses())));
```

Where the unpacking makes it more clear that is where since the square brackets wrap the array bits:
```php
$args = array_unique(array_filter([__CLASS__, $this::class, ...$this->getClasses()]));
```


I also like this one  as i don't have to look at the end of the first line to understand that we are merging, that becomes clear from each of the lines that are being merged:

```php
->will(static::returnValue([
	...array_map([$this, 'createPublicMethod'], $publicMethods),
	...array_map([$this, 'createPrivateMethod'], $privateMethods),
]));
```
vs:
```php
->will(static::returnValue(array_merge(
	array_map([$this, 'createPublicMethod'], $publicMethods),
	array_map([$this, 'createPrivateMethod'], $privateMethods)
)));
```

So to sum it up, for me it's similar to the use of `[]` instead of `array()`. There is no big impact of it it's just new syntax to make things a little bit nicer.

In most cases the lines are shorter, there are few lines, or over all fewer symbols. It's only when you are merging 4 arrays that you would start it would start to get longer then `array_merge()`  and we don't have any such cases.

> Have you considered the varying behavior of both ways?

Yes

> Also, if [this comment](https://www.php.net/manual/en/function.array-merge.php#126687) is true, it might even be a minor performance downgrade.

I ran a few yest and performance was identical.

Running the one from the comment I actually get better performance using spread then `array_merge()`:

```
0.38876485824585 sec for spread
0.43788719177246 sec for array_merge
```

I also added a test for the inline case where the performance is even better:

```php
$before = microtime(true);

for ($i=0 ; $i<10000000 ; $i++) {
    $array2 = ['carrot','lettuce','broccoli'];

    $array1 = ['apple','orange','banana',...$array2];
}

$after = microtime(true);
echo ($after-$before) . " sec for inline spread\n";
```

```
0.36439108848572 sec for inline spread
```
(I'm running PHP 8.3.6 here)

**Comment by AJenbo** [comment]
For completeness:
```
0.39206385612488 sec for spread
0.44558620452881 sec for array_merge
0.36626315116882 sec for inline spread
0.56836605072021 sec for array_push
0.57034206390381 sec for array_unshift
0.60777497291565 sec for foreach
```

I guess we can expect a 11-17% performance increase from this :shrug:
We should also hunt down any place where foreach is being used as it is very slow for combining arrays.

```php
<?php
$before = microtime(true);
for ($i=0 ; $i<10000000 ; $i++) {
    $array1 = ['apple','orange','banana'];
    $array2 = ['carrot','lettuce','broccoli'];

    $array1 = [...$array1,...$array2];
}
$after = microtime(true);
echo ($after-$before) . " sec for spread\n";

$before = microtime(true);
for ($i=0 ; $i<10000000 ; $i++) {
    $array1 = ['apple','orange','banana'];
    $array2 = ['carrot','lettuce','broccoli'];

    $array1 = array_merge($array1,$array2);
}
$after = microtime(true);
echo ($after-$before) . " sec for array_merge\n";

$before = microtime(true);
for ($i=0 ; $i<10000000 ; $i++) {
    $array2 = ['carrot','lettuce','broccoli'];

    $array1 = ['apple','orange','banana',...$array2];
}
$after = microtime(true);
echo ($after-$before) . " sec for inline spread\n";

$before = microtime(true);
for ($i=0 ; $i<10000000 ; $i++) {
    $array1 = ['apple','orange','banana'];
    $array2 = ['carrot','lettuce','broccoli'];

    array_push($array1,...$array2);
}
$after = microtime(true);
echo ($after-$before) . " sec for array_push\n";

$before = microtime(true);
for ($i=0 ; $i<10000000 ; $i++) {
    $array1 = ['apple','orange','banana'];
    $array2 = ['carrot','lettuce','broccoli'];

    array_unshift($array2,$array1);
}
$after = microtime(true);
echo ($after-$before) . " sec for array_unshift\n";

$before = microtime(true);
for ($i=0 ; $i<10000000 ; $i++) {
    $array1 = ['apple','orange','banana'];
    $array2 = ['carrot','lettuce','broccoli'];
	foreach ($array2 as $value) {
		$array1[] = $value;
	}
}
$after = microtime(true);
echo ($after-$before) . " sec for foreach\n";
```

**Comment by AJenbo** [comment]
Older versions via 3v4l.org (and a lower sample size to allow it to finish in the given time limit)

PHP 8.2.19
```php
0.0014340877532959 sec for spread
0.0015029907226562 sec for array_merge
0.0013399124145508 sec for inline spread
0.0018529891967773 sec for array_push
0.0018692016601562 sec for array_unshift
0.0018501281738281 sec for foreach
```

PHP 8.1.28:
```php
0.00092291831970215 sec for spread
0.0010159015655518 sec for array_merge
0.00088095664978027 sec for inline spread
0.0012168884277344 sec for array_push
0.0013630390167236 sec for array_unshift
0.0012340545654297 sec for foreach
```

**Comment by ravage84** [comment]
While I'm sold on the splat operator, I'm not so much on the plus operator (union operation) on arrays.

Reading ``array_merge`` makes the intent clear and it's easier to reach the PHP doc on that one, in case one needs to re-read the behavior specifics versus the ones of some plus sign on arrays.

Also, applying newer, less character containing syntax to make convoluted code more readable is more of a workaround than the way to go for me.

Lastly, I'm not sure we introduce minor changes in behavior in the name of getting a bit more readable code since union and merge isn't the same after all because we might not have some edge cases not covered by tests.

Anyway, I'm eager to hear what the others think about this.

Thank you for the performance tests.

**Comment by AJenbo** [comment]
One thing I prefer from using union over array_merge is that the intent to keep elements or overwrite them becomes explicit. With `array_merge()` it might combine, overwrite, or even a combination of the two depending on the type of keys it detects.

File: src/main/php/PDepend/Source/AST/ASTEnum.php [hunk 1]
```diff
@@ -220,7 +217,7 @@ public function getInterfaces()
             $interfaces[] = $backedEnum;
         }
 
-        return new ASTArtifactList(array_merge($interfaces, $this->getInterfacesClasses()));
+        return new ASTArtifactList([...$interfaces, ...$this->getInterfacesClasses()]);
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Source/AST/ASTEnum.php:220]: Why not using + here?

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php [hunk 2]
```diff
@@ -625,12 +625,12 @@ protected function parseUseDeclarationVersion70(array $fragments): void
 
             // Add mapping between image and qualified name to symbol table
             if ($image !== false) {
-                $this->useSymbolTable->add($image, implode('', array_merge($fragments, $subFragments)));
+                $this->useSymbolTable->add($image, implode('', [...$fragments, ...$subFragments]));
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:628]: No + here? I'm just trying to get when we use one or the other.

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php [hunk 3]
```diff
@@ -625,12 +625,12 @@ protected function parseUseDeclarationVersion70(array $fragments): void
...
             }
         }
 
         if (isset($image, $subFragments) && $image !== false) {
-            $this->useSymbolTable->add($image, implode('', array_merge($fragments, $subFragments)));
+            $this->useSymbolTable->add($image, implode('', [...$fragments, ...$subFragments]));
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:633]: Maybe there too

**Review by kylekatarnls — APPROVED** [review]

File: src/main/php/PDepend/Source/AST/ASTEnum.php (hunk 1 above)
Comment (AJenbo) [reply src/main/php/PDepend/Source/AST/ASTEnum.php:220]: Union (+) add keys that doesn't exist on the left side. Unpacking (...) takes the values from the array and inserts them in the context.
```
[0, 1, 2] + [3, 4, 5, 6] === [0, 1, 2, 6]
```
So when you have an integer key, you likely do not want to use union to combine arrays.

Instead what we want is an array that contains all elements of both array, and we do not care about the keys, which is what unpacking does.
```
[...[0, 1, 2], ...[3, 4, 5, 6]] === [0, 1, 2, 3, 4, 5, 6]
```

`array_merge()` tries to be clever and will take the value of any element that has an integer key, but overwrite anything that has has a string key. It does this regardless. This however doesn't work well in combination with array keys being converted from string to int if it looks like a number.

```
array_merge(['0' => 11, 'b' => 3], ['0' => 13, 'a' => 5]) === [0 => 11, 'b' => 3, 1 => 13, 'a' => 5]
```

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php (hunk 3 above)
Comment (AJenbo) [reply src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:633]: We want to combine all elements, not just the ones with unique keys.
